release/train_tests/benchmark/recsys/criteo.py

```diff
@@ -0,0 +1,237 @@
+            # transpose().reshape(-1) introduces a copy
+            values=torch.from_numpy(sparse.transpose(1, 0).reshape(-1)),
+            lengths=lengths,
+            offsets=offsets,
+            stride=batch_size,
+            length_per_key=length_per_key,
+            offset_per_key=offset_per_key,
+            index_per_key=index_per_key,
+        ),
+        labels=torch.from_numpy(labels.reshape(-1)),
+    )
+
+
+def read_json_from_s3(bucket_name, key):
+    s3 = boto3.client("s3")
+
+    # Download object content
+    response = s3.get_object(Bucket=bucket_name, Key=key)
+    content = response["Body"].read().decode("utf-8")
+
+    # Parse JSON
+    data = json.loads(content)
+    return data
+
+
+def get_ray_dataset(stage: str = "train"):
+    ctx = DataContext.get_current()
+    # Enable hash-shuffling
+    ctx.shuffle_strategy = ShuffleStrategy.HASH_SHUFFLE
+
+    stage_to_path = {
+        # TODO: The preprocessor fitting (groupby().count()) takes too long
+        # on the full training dataset, so just run with the valid dataset for now.
+        "train": "s3://ray-benchmark-data-internal/criteo/tsv.gz/valid",
+        # "train": "s3://ray-benchmark-data-internal/criteo/tsv.gz/train/",
+        "valid": "s3://ray-benchmark-data-internal/criteo/tsv.gz/valid/",
+        "test": "s3://ray-benchmark-data-internal/criteo/tsv.gz/test/",
+    }
+    ds_path = stage_to_path[stage]
+
+    ds = ray.data.read_csv(
+        ds_path,
+        read_options=pyarrow.csv.ReadOptions(column_names=DEFAULT_COLUMN_NAMES),
+        parse_options=pyarrow.csv.ParseOptions(delimiter="\t"),
+        shuffle=("files" if stage == "train" else None),  # coarse file-level shuffle
+    )
+
+    # Convert categorical features to integers.
+
+    # Option 1: Use the built-in preprocessor (This is not performant enough.)
+    # from ray.data.preprocessors import OrdinalEncoder
+    # categorical_to_indices_preprocessor = OrdinalEncoder(columns=DEFAULT_CAT_NAMES)
+    # ds = categorical_to_indices_preprocessor.fit_transform(ds)
+
+    # Option 2: groupby() w/ hash-based shuffle
+
+    # Fetch cached value counts instead of "fitting" the preprocessor from scratch.
+    COMPUTE_VALUE_COUNTS_FROM_SCRATCH: bool = False
+
+    FREQUENCY_THRESHOLD = 3
+    LOW_FREQUENCY_INDEX = 1  # map low frequency values -> 1
+    categorical_value_to_index = {}
+    for cat_feature in DEFAULT_CAT_NAMES:
+        if COMPUTE_VALUE_COUNTS_FROM_SCRATCH:
+            logger.info(f"Computing value counts for: {cat_feature}")
+
+            def filter_features(batch):
+                features = batch[cat_feature]
+                features[features == None] = ""
+                return {cat_feature: features}
+
+            value_counts = [
+                (group[cat_feature], group["count()"])
+                for group in ds.map_batches(filter_features)
+                .groupby(key=cat_feature)
+                .count()
+                .take_all()
+            ]
```

Comment:
Right now I'm blocked on this taking too long on the full TB dataset, which is why I'm just setting the training set to be the smaller validation set for now. I precomputed these values and saved them as files on s3.

Reply:
Do you think in practice people would want to do this on the fly, instead of precomputing it? 
It looks that there is no randomization, so probably not necessary to repeatedly do this for each epoch.